### PR TITLE
fix(Scripts/Quest): Make quest "Triage" (6622/6624) completable

### LINKED ISSUE #16161: [First Aid] Triage quest problem - patients die too fast
### What client do you play on?

enGB

### Faction

Both

### Content Phase:

Generic

### Current Behaviour

Quests: https://wowgaming.altervista.org/aowow/?quest=6624 for Ally
https://wowgaming.altervista.org/aowow/?quest=6622 for Horde.

Respawned patients died too fast - it's not possible to heal any kinds of patients with quest item.
Video: https://youtu.be/X_OhBaQcQ58

### Expected Blizzlike Behaviour

   It's best to position yourself just beyond the trauma surgeon, between him and the bunk beds, before you take the quest. Also have a spare hotkeyed button available for the bandages he gives you, so you can apply them faster.
    When you take the quest, the surgeon will give you [Triage Bandage] (20 charges). After taking the bandages, drag them from your backpack to the button you set aside as fast as possible and look for injured soldiers on the bunk beds. If you are positioned right, you can bandage them without moving or by only moving a little.
    Bandage the patients in this priority order:
        Critically Injured
        Badly Injured
        Injured
        If there are multiple patients of the same type, try to bandage the first one that you remember appearing, because that one will probably fade first.
    Click on the patient and immediately hit the hotkey for your bandage. While you are bandaging, look around (try holding down left-mouse-button and move the mouse around) for the next patient to bandage in priority order.
    You can select a new patient even while still bandaging the current patient, which will save you a lot of time. As soon as bandaging completes, hit the key for your bandage again to begin bandaging this new patient. If you have to move, try to do it quickly (click both mouse buttons while looking at the patient).
    Keep doing steps 5 and 6 while remembering the priority in step 4. Eventually the surgeon will show a gold Activequest and you're done. You should have 5 bandages left.

That's pretty much the system. You will lose a few patients (usually 2-4), but should complete the quest with no problems. 

### Source

https://www.youtube.com/watch?v=CXM162uXIjo
https://www.youtube.com/watch?v=TdMOERz3pAs
https://wowpedia.fandom.com/wiki/Triage/Artisan_First_Aid_quest_guide

### Steps to reproduce the problem

.learn 50299
.setskill 129 450

Alliance
.go c 33780
Horde
.go c 14250

Accept quest
Look at victims

### Extra Notes

https://github.com/chromiecraft/chromiecraft/issues/5507

### AC rev. hash/commit

https://github.com/chromiecraft/azerothcore-wotlk/commit/ce6ca597b54e64f328cacddae5f003cdc41a343d

### Operating system

Ubuntu 20.04

### Modules

- [mod-ah-bot](https://github.com/azerothcore/mod-ah-bot)
- [mod-bg-item-reward](https://github.com/azerothcore/mod-bg-item-reward)
- [mod-BreakingNewsOverride](https://github.com/AnchyDev/BreakingNewsOverride)
- [mod-cfbg](https://github.com/azerothcore/mod-cfbg)
- [mod-chat-transmitter](https://github.com/azerothcore/mod-chat-transmitter)
- [mod-cta-switch](https://github.com/azerothcore/mod-cta-switch)
- [mod-desertion-warnings](https://github.com/azerothcore/mod-desertion-warnings)
- [mod-duel-reset](https://github.com/azerothcore/mod-duel-reset)
- [mod-eluna](https://github.com/azerothcore/mod-eluna)
- [mod-instanced-worldbosses](https://github.com/nyeriah/mod-instanced-worldbosses)
- [mod-ip-tracker](https://github.com/azerothcore/mod-ip-tracker)
- [mod-low-level-arena](https://github.com/azerothcore/mod-low-level-arena)
- [mod-low-level-rbg](https://github.com/azerothcore/mod-low-level-rbg)
- [mod-multi-client-check](https://github.com/azerothcore/mod-multi-client-check)
- [mod-progression-system](https://github.com/azerothcore/mod-progression-system)
- [mod-pvp-titles](https://github.com/azerothcore/mod-pvp-titles)
- [mod-pvpstats-announcer](https://github.com/azerothcore/mod-pvpstats-announcer)
- [mod-queue-list-cache](https://github.com/azerothcore/mod-queue-list-cache)
- [mod-rdf-expansion](https://github.com/azerothcore/mod-rdf-expansion)
- [mod-transmog](https://github.com/azerothcore/mod-transmog)
- [mod-weekend-xp](https://github.com/azerothcore/mod-weekend-xp)
- [mod-zone-difficulty](https://github.com/azerothcore/mod-zone-difficulty)
- [lua-carbon-copy](https://github.com/55Honey/Acore_CarbonCopy)
- [lua-event-scripts](https://github.com/55Honey/Acore_eventScripts)
- [lua-exchange-npc](https://github.com/55Honey/Acore_ExchangeNpc)
- [lua-level-up-reward](https://github.com/55Honey/Acore_LevelUpReward)
- [lua-recruit-a-friend](https://github.com/55Honey/Acore_RecruitAFriend)
- [lua-send-and-bind](https://github.com/55Honey/Acore_SendAndBind)
- [lua-temp-announcements](https://github.com/55Honey/Acore_TempAnnouncements)
- [lua-zonecheck](https://github.com/55Honey/acore_Zonecheck)

### Customizations

None

### Server

ChromieCraft PTR - AzerothCore rev. df77b5f4f79d 2023-04-24 08:13:07 -0300 (master branch) (Unix, Debug, Static)

### PR DESCRIPTION
**Pull request by kjack9**
## Changes Proposed:
- removes tick-based health reduction to patients and replaces with an emulation of the originally-intended effect
- fills all patient beds immediately to match Blizzlike behavior
- continue summoning patients long enough to complete the quest objective
- don't kill the last patient saved
- make patients invulnerable to non-quest-item healing

## Issues Addressed:

- Closes #16161
- Closes chromiecraft/chromiecraft#5507

## SOURCE:

- personal questing

## Tests Performed:

- completed Alliance version successfully multiple times
- completed Horde version

## How to Test the Changes:

1. Have First Aid skill of at least 225
2. Go to Doctor Gustaf VanHowzen (Alliance, 12939) or Doctor Gregory Victor (Horde, 12920)
3. Start quest, observe patients in beds
4. Use quest item to heal patients
5. Note completed quest

## Known Issues and TODO List:

- True BlizzLike would have each of the patients being replaced independently every 15-30 seconds. Currently the doctor is actually doing the placing and the script is not setup for this. Maybe summon using events instead of a timer variable?

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by kjack9** [comment]
This is my first crack at updating an AI script - please let me know what I can be doing better for next time.

File: src/server/scripts/World/npcs_special.cpp [hunk 1]
```diff
@@ -990,7 +990,7 @@ class npc_doctor : public CreatureScript
                 Reset();
         }
 
-        void PatientSaved(Creature* /*soldier*/, Player* player, Location* point)
+        void PatientSaved(Creature* saved_patient, Player* player, Location* point)
```
Comment (Kitzunu) [inline src/server/scripts/World/npcs_special.cpp:993]:
```suggestion
        void PatientSaved(Creature* savedPatient, Player* player, Location* point)
```

File: src/server/scripts/World/npcs_special.cpp [hunk 2]
```diff
@@ -1090,6 +1094,14 @@ class npc_injured_patient : public CreatureScript
                     me->SetHealth(me->CountPctFromMaxHealth(25));
                     break;
             }
+
+            // emulate vanilla WoW's 200ms spell batching rate for maximum authenticity!
+            // https://www.wowhead.com/classic/news/spell-batching-in-classic-wow-290910
```
Comment (Kitzunu) [inline src/server/scripts/World/npcs_special.cpp:1099]: Okay, but the question is. Did this exist in 3.3.5?

File: src/server/scripts/World/npcs_special.cpp (hunk 1 above)
Comment (kjack9) [reply src/server/scripts/World/npcs_special.cpp:993]: Sorry, my Python is showing. I'll fix it.

File: src/server/scripts/World/npcs_special.cpp (hunk 2 above)
Comment (kjack9) [reply src/server/scripts/World/npcs_special.cpp:1099]: So, I don't currently have a character in WotLK Classic to test with that has First Aid 225. All the videos I can find have the person _successfully_ completing the quest, which means that not enough guys die to tell how the timing works. Worse, the live version doesn't give you decreasing healthbars - the soldiers all have "full" health and then up and die after a set timer.

Given that I didn't have a way to find out for sure, I just replicated the existing functionality that relied on the vanilla tick rate of every 200ms, presuming that whoever wrote that 7+ years ago was about right for timing. I figure working and close is better than utterly broken.

If anyone has the inclination, a video of the quest getting started and then just waiting while all the soldiers die (with them in view) would be really helpful in replicating the actual quest functionality.

Comment (kjack9) [reply src/server/scripts/World/npcs_special.cpp]: > Okay, but the question is. Did this exist in 3.3.5?

Dang it, you made me want to figure this out. I power-leveled first aid for a character on WotLK Classic and figured out the exact timing.

- Critically Injured: dies after 25s
- Badly Injured: dies after 35s
- Injured: dies after 65s

Each of the soldiers spawns in a range of levels, so the amount of actual health is not consistent. So I am simply setting the starting health to the number of seconds needed (25/35/65%) and then reducing their health by 1% per second until they are dead. Simple and replicates WotLK Classic pretty much exactly. Without access to a real 3.3.5 server, I think this is as Blizzlike as I can make the dying time for each solider type.

**Comment by kjack9** [comment]
Tested the Horde version (12920) too, working fine.

**Comment by kjack9** [comment]
Any outstanding concerns or questions about this change?

**Review by Nyeriah — APPROVED** [review]

**Pull request merged**